api/v1alpha1/addon_types.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 D2iQ, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package v1alpha1
+
+import (
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	"github.com/d2iq-labs/capi-runtime-extensions/common/pkg/capi/clustertopology/variables"
+)
+
+type Addons struct {
+	// +optional
+	CNI *CNI `json:"cni,omitempty"`
+
+	// +optional
+	NFD *NFD `json:"nfd,omitempty"`
+
+	// +optional
+	CPI *CPI `json:"cpi,omitempty"`
+
+	// +optional
+	CSIProviders *CSIProviders `json:"csi,omitempty"`
+}
+
+func (Addons) VariableSchema() clusterv1.VariableSchema {
+	return clusterv1.VariableSchema{
+		OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+			Description: "Cluster configuration",
+			Type:        "object",
+			Properties: map[string]clusterv1.JSONSchemaProps{
+				"cni": CNI{}.VariableSchema().OpenAPIV3Schema,
+				"nfd": NFD{}.VariableSchema().OpenAPIV3Schema,
+				"csi": CSIProviders{}.VariableSchema().OpenAPIV3Schema,
+				"cpi": CPI{}.VariableSchema().OpenAPIV3Schema,
+			},
+		},
+	}
+}
+
+type AddonStrategy string
+
+const (
+	AddonStrategyClusterResourceSet AddonStrategy = "ClusterResourceSet"
+)
+
+// CNI required for providing CNI configuration.
+type CNI struct {
+	// +optional
+	Provider string `json:"provider,omitempty"`
+	// +optional
+	Strategy AddonStrategy `json:"strategy,omitempty"`
```

Comment:
Remind me again please how does `optional` and `omitempty` work here with `Required` on line 75. Should these not be marked optional here?

Reply:
Optional and omitempty in these API definitions are standard when a default can be added via mutating webhook - standard k8s API behaviour. While that might seem odd that the OpenAPI definition says it is required, that is OK because the OpenAPI has a default declared. I agree it is a bit confusing, I think this is actually OK. If the OpenAPI definition was not set as required then `strategy: null` would be acceptable and not defaulted (at least I think that's the expected behaviour!).